nixos/modules/services/home-automation/esphome.nix

```diff
@@ -0,0 +1,153 @@
+{ config, lib, pkgs, ... }:
+
+let
+  inherit (lib)
+    literalExpression
+    maintainers
+    mkEnableOption
+    mkIf
+    mkOption
+    mdDoc
+    types
+    ;
+
+  cfg = config.services.esphome;
+
+  stateDir = "/var/lib/esphome";
+
+  esphomeParams =
+    if cfg.enableUnixSocket
+    then "--socket /run/esphome/esphome.sock"
+    else "--address ${cfg.address} --port ${toString cfg.port}";
+in
+{
+  meta.maintainers = with maintainers; [ oddlama ];
+
+  options.services.esphome = {
+    enable = mkEnableOption (mdDoc "esphome");
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.esphome;
+      defaultText = literalExpression "pkgs.esphome";
+      description = mdDoc "The package to use for the esphome command.";
+    };
+
+    enableUnixSocket = mkOption {
+      type = types.bool;
+      default = false;
+      description = lib.mdDoc "Listen on a unix socket `/run/esphome/esphome.sock` instead of the TCP port.";
+    };
+
+    address = mkOption {
+      type = types.str;
+      default = "localhost";
+      description = mdDoc "esphome address";
+    };
+
+    port = mkOption {
+      type = types.port;
+      default = 6052;
+      description = mdDoc "esphome port";
+    };
+
+    openFirewall = mkOption {
+      default = false;
+      type = types.bool;
+      description = mdDoc "Whether to open the firewall for the specified port.";
+    };
+
+    allowedDevices = mkOption {
+      default = [];
+      example = [{ node = "/dev/serial/by-id/usb-Silicon_Labs_CP2102_USB_to_UART_Bridge_Controller_0001-if00-port0"; modifier = "rw"; }];
+      description = lib.mdDoc ''
+        A list of device nodes to which {command}`esphome` has access to.
+        Beware that permissions are not added dynamically when a device
+        is plugged in while the service is already running.
+      '';
+      type = types.listOf (types.submodule {
+        options = {
+          node = mkOption {
+            example = "/dev/ttyUSB*";
+            type = types.str;
+            description = lib.mdDoc "Path to device node";
+          };
+          modifier = mkOption {
+            example = "rw";
+            type = types.str;
+            description = lib.mdDoc ''
+              Device node access modifier. Takes a combination
+              `r` (read), `w` (write), and `m` (mknod). See the
+              `systemd.resource-control(5)` man page for more
+              information.
```

Comment:
I originally just adopted this from the nix-containers definition, since they also needed this. But you are right, in this case it always makes sense to use rw. Will change.